### PR TITLE
Show an error page if the rest api is not available

### LINKED ISSUE #1389: Show an error page if the REST API is not available
**Is your feature request related to a problem? Please describe.**
When the rest endpoint is not reachable due to a problem, the angular application doesn't work but show only a page with an hanging load spinner. It'd be nice instead to have a 500 page where to redirect in this case

**Describe the solution you'd like**
A possible solution is to create an angular guard that make a rest request to `/server/api` and in case it gets an error response it redirect to a general 500 page (that looks like the 404 page)

### PR DESCRIPTION
**Pull request by atarix83**
## References
* Fixes #1389 

## Description
This PR add a fallback strategy on angular application when the rest server api doesn't work. Indeed an error page is automatically shown when rest server is not reachable

Schermata da 2021-11-17 09-32-03

## Instructions for Reviewers
For testing this PR it's sufficient to change the angular environment file to point to an wrong rest server host. It's possible to test also cors errors by changing the ui running port so that it is different from 4000

List of changes in this PR:
* Created a new 500 error page
* Created a `ServerCheckGuard` that verify if root api endpoint works properly, if not redirect to the 500 error page

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Review by tdonohue — CHANGES_REQUESTED** [review]
@atarix83 : I tested this today & it's not fully working when I use `yarn start`.  

Here's what I tried:
1. Start frontend with `yarn start`
2. Stop backend entirely
3. Go to http://localhost:4000/ in browser

What I see is that the "Service Unavailable" page briefly flashes, but then it goes to an empty page with the moving dots (i.e. the loading screen) like this:

homepage

In the commandline where I ran `yarn start`, I also see two errors appear:
```
ERROR Error: undefined doesn't contain the link sites
    at MapSubscriber.project (C:\programming\dspace-angular\dist\server\main.js:1:7292256)
    at MapSubscriber._next (C:\programming\dspace-angular\dist\server\main.js:1:6083990)
    at MapSubscriber.next (C:\programming\dspace-angular\dist\server\main.js:1:1152484)
    at MapSubscriber._next (C:\programming\dspace-angular\dist\server\main.js:1:6084096)
    at MapSubscriber.next (C:\programming\dspace-angular\dist\server\main.js:1:1152484)
    at TakeSubscriber._next (C:\programming\dspace-angular\dist\server\main.js:1:2788386)
    at TakeSubscriber.next (C:\programming\dspace-angular\dist\server\main.js:1:1152484)
    at FilterSubscriber._next (C:\programming\dspace-angular\dist\server\main.js:1:6587868)
    at FilterSubscriber.next (C:\programming\dspace-angular\dist\server\main.js:1:1152484)
    at MapSubscriber._next (C:\programming\dspace-angular\dist\server\main.js:1:6084096)
ERROR TypeError: authorizations.filter is not a function
    at SwitchMapSubscriber.project (C:\programming\dspace-angular\dist\server\main.js:1:412900)
    at SwitchMapSubscriber._next (C:\programming\dspace-angular\dist\server\main.js:1:5083453)
    at SwitchMapSubscriber.next (C:\programming\dspace-angular\dist\server\main.js:1:1152484)
    at CatchSubscriber.notifyNext (C:\programming\dspace-angular\dist\server\main.js:1:8802928)
    at SimpleInnerSubscriber._next (C:\programming\dspace-angular\dist\server\main.js:1:8802623)
    at SimpleInnerSubscriber.next (C:\programming\dspace-angular\dist\server\main.js:1:1152484)
    at Observable_Observable._subscribe (C:\programming\dspace-angular\dist\server\main.js:1:6343896)
    at Observable_Observable._trySubscribe (C:\programming\dspace-angular\dist\server\main.js:1:2650831)
    at Observable_Observable.subscribe (C:\programming\dspace-angular\dist\server\main.js:1:2650613)
    at innerSubscribe (C:\programming\dspace-angular\dist\server\main.js:1:8803238)
```

However, if I stop the UI and restart it using `yarn start:dev`, then everything works as expected.

**Comment by tdonohue** [comment]
@atarix83 : It looks like this PR now has a merge conflict...which I believe was caused by the Angular 11 upgrade in #1403 .  Could you quickly rebase this on the latest `main`?  I'm hoping the conflict is relatively minor

**Review by tdonohue — COMMENTED** [review]
@atarix83 : Tested again, and this now works mostly as described.  I have one very minor issue noted below (unused or unnecessary i18n key).

However, I did want to note two minor usability issues that I've encountered (when running `yarn start`). We could choose to delay these for a different PR though, and simply move forward with this PR as-is.

1. After encountering the 500 error, the UI never refreshes itself when the REST API is available again.  Here's how to reproduce:
    * Stop the REST API
    * Access http://localhost:4000/ -> redirected to http://localhost:4000/500 and the error message is properly displayed
    * Start the REST API
    * Back in your browser, click the DSpace icon in the header, or click a header menu.  You will remain on the http://localhost:4000/500 page, will continue to see the "Service Unavailable" error message.  (I'd expect for the page to now detect that the REST API is available and refresh/reload itself.)
    * The only way to get the site to work again is to click the reload button in your browser, and then it will work as expected.
2. If REST API crashes after you have loaded the UI, then the UI will just "hang" (similar to `main` branch behavior).  Here's how to reproduce:
    * Make sure REST API is running.
    * Access http://localhost:4000/
    * Stop the REST API
    * Back in your browser, type in a search term and click "Search" button.  Or, click on a link in the header (e.g. "Communities & Collections")
    * In either scenario, you will just see the "Loading..." message, and never be redirected to the new `/500` page
    * If you now click the reload button in your browser window, the `/500` page will appear.
    
    
Overall, I think this current PR is much better behavior than what we have on `main`.  So, if you'd rather move this forward as-is, I'm OK with approving this PR.  I just wanted to point out those two usability issues that remain with this feature.

File: src/assets/i18n/en.json5 [hunk 1]
```diff
@@ -14,6 +14,11 @@
 
   "403.forbidden": "forbidden",
 
+  "500.page-internal-server-error": "Service Unavailable",
+
+  "500.help": "The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later.",
+
+  "500.link.home-page": "Take me to the home page",
```
[the remark this answers is not in the document]
Comment (atarix83) [reply src/assets/i18n/en.json5:21]: @tdonohue 

all the points should be resolved now, could you test it again

**Comment by tdonohue** [comment]
@atarix83 : This PR appears to have merge conflicts now.  Could you quickly rebase on main?

**Review by tdonohue — COMMENTED** [review]
@atarix83 : The previous issues are now resolved & this is working as expected!  However, it seems there are now some e2e test failures specific to this PR (they don't appear on `main` and I've tried rerunning the e2e tests, but they consistently fail now)

**Comment by tdonohue** [comment]
@atarix83 : I looked into why this is failing some e2e tests...and it seems that **somehow** this PR (occasionally?) "breaks" (or hides) the "Statistics" link in the header...at least when you run it using `yarn start`.  If I run this PR, and reload the homepage several times (just by clicking the browser's reload button), sometimes the Statistics link appears, and sometimes it does NOT.  Here's a screenshot from one of our e2e tests where it's missing.
**Screenshot of failing e2e Community Statistics test (no "Statistics" link in header)**
Community Statistics Page -- should load if you click on Statistics from a Community page (failed)


There's no errors appearing in my DevTools Console when this occurs, but it's occurring consistently.  If you look closely at the failing e2e tests, around half of them relate to Statistics, and they are failing as they timeout waiting for that link to appear in the header.

The other failing e2e tests all look to be timeouts as the pages have not fully loaded. This is based on the video/image artifacts you can download at the bottom of the [Test summary page](https://github.com/DSpace/dspace-angular/actions/runs/1583044329)... when the failure occurs in the other tests, the screenshot always looks similar to this..

**Screenshot of test timeout...the page is still loading**
Browse By Date Issued -- should pass accessibility tests (failed)

It's not yet clear to me if this PR is somehow causing pages to load slightly slower...or if this timeout is a "random failure".

**Comment by atarix83** [comment]
@tdonohue 

I should have fixed the issue, but now the build fails because it has a problem during docker init. I've already tried to run again the build but nothing changed.

File: src/app/root/root.component.ts [hunk 2]
```diff
@@ -71,5 +71,9 @@ export class RootComponent implements OnInit {
       .pipe(
         map(([collapsed, mobile]) => collapsed || mobile)
       );
+
+    if (this.router.url === '/500') {
```
Comment (artlowel) [inline src/app/root/root.component.ts:75]: Perhaps use `getPageInternalServerErrorRoute()` here instead of repeating '/500'?

**Review by artlowel — CHANGES_REQUESTED** [review]
Thanks @atarix83!

This is a definite improvement. However, currently the actual status code for the `/500` route is 200. You can verify this by disabling javascript and checking the GET of the page in the network tools.

Please set that to 500 as well. We already do something similar for the 4xx pages: https://github.com/DSpace/dspace-angular/blob/84840678d2b725a49b793ba751c3b7147a47c1e0/src/app/core/services/server-response.service.ts

**Review by tdonohue — APPROVED** [review]
👍 Thanks @atarix83 !  Re-tested this today, and it works great for me.  I **was able to reproduce** the issue that @artlowel noted... the `/500` page returns a 200 response code instead of a 500.  I agree that should be fixed before we merge this.

That said, I'll go ahead and add a +1 now.   That way as soon as it's resolved & @artlowel can retest, it's ready to merge.

**Comment by atarix83** [comment]
@artlowel 

feedback addressed

**Review by artlowel — APPROVED** [review]
Thanks @atarix83!

**Comment by tdonohue** [comment]
@atarix83 : Unfortunately, it appears this PR somehow has accidentally broken Authentication when running in Production mode. It appears both on demo7.dspace.org and also I can reproduce locally.  If I revert to `main` before this PR was merged, then everything works.

As best I can tell, it almost looks like something in SSR is failing silently.  The login seems to succeed (REST returns 200), but during the redirect/reload of the homepage, the Angular app just hangs.  Could you possibly look into this tomorrow?  Otherwise, I may need to temporarily revert this PR until we can track down a way to fix it.

**Comment by tdonohue** [comment]
This PR was reverted in #1493.  The original issue ticket in #1389 was reopened

**Pull request merged**